### ISSUE TITLE
ETX on UART causes KeyboardInterrupt

### ISSUE THREAD

**Issue by GreenSlime96** [bbcmicrobit/micropython#423]
Hi all,

I am developing a protocol on top of UART, and right now whenever the character ETX (ASCII 0x03) is sent, the REPL seems to activate after receiving a KeyboardInterrupt.

I am unable to catch the keyboard interrupt and it seems as if there's nothing I can do as the REPL is still running in the background.

Probably a bug as this issue doesn't happen on my PyBoards.

See also https://forum.micropython.org/viewtopic.php?t=2848

Many thanks,
Boon.

**Comment by whaleygeek** [comment]
Check with @ntoll about using the raw REPL as he uses that for binary file
transfer in the Mu editor

On 2 Apr 2017 2:32 am, "Kheng Boon Pek" <notifications@github.com> wrote:

Hi all,

I am developing a protocol on top of UART, and right now whenever the
character ETX (ASCII 0x03) is sent, the REPL seems to activate after
receiving a KeyboardInterrupt.

I am unable to catch the keyboard interrupt and it seems as if there's
nothing I can do as the REPL is still running in the background.

Probably a bug as this issue doesn't happen on my PyBoards.

Many thanks,
Boon.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bbcmicrobit/micropython/issues/423>, or mute the thread

**Comment by dpgeorge** [comment]
There is currently no way around this: the underlying UART driver looks for a ctrl-C character in the UART interrupt handler (whenever a new character comes in) and converts it into a KeyboardInterrupt.

To make it possible to let these characters through would require a new method in the UART class to disable this behaviour, eg (as pyboard does it with USB VCP):
```python
uart.set_interrupt(-1) # don't capture ctrl-C characters
uart.read(10)
# etc
```
The pitfall with this is that you won't be able to interrupt a running script if you disable ctrl-C, but that's just something the users need to understand when using this advanced feature.

**Comment by GreenSlime96** [comment]
Thank you for the clarification @dpgeorge, I looked through the documentation and source and wasn't able to find anything too helpful :)

**Comment by dpgeorge** [comment]
This was fixed upstream: the micropython.kbd_intr() function was added.  But it will be some time before it filters down into this repo.

**Comment by dpgeorge** [comment]
micropython.kbd_intr() was made available in 9af7add02c2769c2fc8c34caff782319e2c19661
